janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexHasIdOptimizerStrategy.java

```diff
@@ -73,4 +77,10 @@ protected boolean isValidPredicate(P<?> predicate) {
         Object predicateValue = predicate.getValue();
         return predicateValue instanceof Vertex || predicateValue instanceof Long;
     }
+
+    @Override
+    protected FilterStep<Edge> makeFilterByAdjacentIdStep(Traversal.Admin<?, ?> traversal, HasStep<?> originalStep) {
+        HasContainer hc = new HasContainer(ADJACENT_ID.name(), P.eq(parsePredicate(originalStep).getValue()));
+        return new HasStep<Edge>(traversal, hc);
```

Comment:
Labels are not needed in this case because the ID is unique even across labels. What we do need is the edge label, but this is already contained in the preceeding `in()`/`out()` step.
Your comment howewer pointed me to the [edge layout](https://docs.janusgraph.org/advanced-topics/data-model/#individual-edge-layout) docs page where I realized we should check if there is a sort key defined for the given edge label before optimizing. Because as one can see, only when the sort key is empty, the sort order will be defined by the IDs.

Reply:
Oh, I meant by label() step in gremlin query, not vertex label or edge label. 😉

Reply:
Can you give an example of a query where you think a `label()` step would be relevant? I can't think of one

Reply:
Regarding the sort key, applying the optimization is still advantageouos even with sort keys defined. This may cause a linear search across the edges for the one with the correct ID but this if of course still faster than a linear search across the neighbors with the correct property value.

Reply:
whoops sorry, I meant by as(stepLabel) step, e.g. `g.V(vertex).out().has("key", "value").as("v").addE("connects").from("v").next()`

Reply:
Finally coming back to this. Very interesting thought, but I think as the only thing that we actually replace are `has` steps, step labels should not interfere. The [TinkerPop documentation](https://tinkerpop.apache.org/docs/current/reference/#has-step) shows various usages of the `has` step. Most of them use only literal values which cannot be derived from step labels. The only exceptions are:
- `has(key, predicate)` which is ok because predicates also only take literal values.
- `has(key, traversal)` which is also ok because nested traversals don't have access to the surrounding traversal's step labels in the original query either.

Reply:
I am not sure if I got you, but let me explain what I am worried about:

Let's say we have a normal query:
```groovy
g.V(v1).as("v1").out().has("prop", "value2").as("v2").select("v1", "v2").by("prop")
```

I suspect that with your change, it would be translated to something equivalent to  (note the missing 'as("v2")' step):
```groovy
g.V(v1).as("v1").out().hasId(v2).select("v1", "v2").by("prop")
```
which leads to the wrong result.

Reply:
Yes, you were right. What I thought about was the impact of labels for the `has` step not _on_ the `has` step. I wrote a test to verify this and indeed had to fix two lines. 
Glad you found that because two existing strategies were also affected.

Reply:
Do you mean the same problem exists in two strategies released in v0.5? If that's the case, I suggest you open up a separate PR to fix them, in case this PR cannot be merged in the next release.

Reply:
Sounds like the right thing to do